transaction/std/src/transaction_builder.rs

```diff
@@ -16,32 +17,43 @@ use mc_transaction_core::{
     onetime_keys::create_shared_secret,
     ring_signature::SignatureRctBulletproofs,
     tx::{Tx, TxIn, TxOut, TxOutConfirmationNumber, TxPrefix},
-    CompressedCommitment,
+    CompressedCommitment, MemoContext, MemoPayload, NewMemoError,
 };
 use mc_util_from_random::FromRandom;
 use rand_core::{CryptoRng, RngCore};
-use std::collections::HashSet;
 
-/// Helper utility for building and signing a CryptoNote-style transaction.
+/// Helper utility for building and signing a CryptoNote-style transaction,
+/// and attaching fog hint and memos as appropriate.
 ///
-/// Note: This is generic over FogPubkeyResolver, because otherwise, to test
-/// this object, we must create fake IAS reports which is quite difficult.
+/// This is generic over FogPubkeyResolver because there are several reasonable
+/// implementations of that.
+///
+/// This is generic over MemoBuilder to allow injecting a policy for how to
+/// use the memos in the TxOuts.
 #[derive(Debug)]
 pub struct TransactionBuilder<FPR: FogPubkeyResolver> {
     /// The input credentials used to form the transaction
     input_credentials: Vec<InputCredentials>,
     /// The outputs created by the transaction, and associated shared secrets
     outputs_and_shared_secrets: Vec<(TxOut, RistrettoPublic)>,
-    /// The tombstone_block value, a block index after which the transaction
-    /// expires.
+    /// The tombstone_block value, a block index in which the transaction
+    /// expires, and can no longer be added to the blockchain
```

Comment:
Is this a documentation bug fix?

Reply:
that was what i was trying to do

fog is currently assuming that it doesn't have to scan the `pubkey-expiry` block, because people told me that a transaction is dead in the tombstone block. (And tombstone-block <= pubkey-expiry is an invariant we enforce in the transaction builder).

However, in the consensus code what I'm seeing is:

```
        // find the expired entries and remove them, storing their keys in expired,
        // without destroying or re-allocating the cache
        cache.retain(|key, entry| -> bool {
            if entry.context().tombstone_block() < block_index {
                expired.insert(*key);
                false
            } else {
                true
            }
        });
```

So maybe there is a mismatch in definitions here.

What I would like to see is to change the byzantine worker code to:

```
        // find the expired entries and remove them, storing their keys in expired,
        // without destroying or re-allocating the cache
        cache.retain(|key, entry| -> bool {
            if entry.context().tombstone_block() <= block_index {
                expired.insert(*key);
                false
            } else {
                true
            }
        });
```

Otherwise we will have to change a lot of fog-view and fog database code to allow that a transaction can use a fog key in the block in which it expires

Reply:
It looks that in the actual transaction validation code, we check on the tombstone block, and we have defined that the transaction is invalid if the tombstone_block is <= the current block index: https://github.com/mobilecoinfoundation/mobilecoin/blob/103aa92bb0c8934d26de01b16c2ef1cbeebe3c95/transaction/core/src/validation/validate.rs#L358

So I think this is indeed a documentation bug fix, regardless of what's in `TxManager`